File: templates/hook-template.py
from dataclasses import dataclass
from typing import Generic, Literal, Protocol, TypeVar

T = TypeVar("T")

@dataclass(frozen=True)
class HookContext:
    tenant_id: str
    user_id: str
    session_id: str
    run_id: str

@dataclass
class HookResult(Generic[T]):
    action: Literal["continue", "deny", "annotate"]
    value: T
    reason: str | None = None

class AgentHook(Protocol[T]):
    name: str
    event: str
    priority: int
    timeout_seconds: int
    failure_mode: Literal["closed", "open"]
    def run(self, value: T, context: HookContext) -> HookResult[T]: ...

File: templates/hook-template.ts
export type HookEvent = "PreSessionStart" | "PostSessionStart" | "PreContextBuild" | "PostContextBuild" | "PreModelCall" | "PostModelCall" | "PreToolUse" | "PostToolUse" | "OnToolError" | "PreCompact" | "PostCompact" | "PreFinalAnswer" | "PostFinalAnswer" | "OnSessionEnd";
export interface HookContext { tenantId: string; userId: string; sessionId: string; runId: string; signal: AbortSignal; }
export interface HookResult<T> { action: "continue" | "deny" | "annotate"; value: T; reason?: string; }
export interface AgentHook<T> { name: string; event: HookEvent; priority: number; timeoutMs: number; failureMode: "closed" | "open"; run(value: T, context: HookContext): Promise<HookResult<T>>; }
